src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,95 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
+#nullable enable
 
 #if NETCOREAPP
 
+using System;
+using System.Collections.Generic;
 using System.Diagnostics;
+using System.IO;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.Loader;
 
 namespace Microsoft.CodeAnalysis
 {
     internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
     {
-        private AssemblyLoadContext _loadContext;
+        private readonly object _guard = new object();
+        private readonly Dictionary<string, AssemblyLoadContext> _loadContextByDirectory = new Dictionary<string, AssemblyLoadContext>();
 
         protected override Assembly LoadFromPathImpl(string fullPath)
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
-            {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+            AssemblyLoadContext? loadContext;
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException();
+            lock (_guard)
+            {
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
                 {
-                    _loadContext.Resolving += (context, name) =>
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath);
+                    loadContext.Resolving += (context, name) =>
                     {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
+                        Debug.Assert(ReferenceEquals(context, loadContext));
                         return Load(name.FullName);
                     };
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
                 }
             }
 
-            return LoadImpl(fullPath);
+            return loadContext.LoadFromAssemblyPath(fullPath);
         }
 
-        protected virtual Assembly LoadImpl(string fullPath) => _loadContext.LoadFromAssemblyPath(fullPath);
+        private class DirectoryLoadContext : AssemblyLoadContext
+        {
+            private readonly string _directory;
+            public DirectoryLoadContext(string fullDirectoryPath)
+            {
+                _directory = fullDirectoryPath;
+            }
+
+            protected override Assembly? Load(AssemblyName assemblyName)
```

Comment:
Overall this doesn't feel correct because the compiler is respecting the runtime state of the files in `_directory`. That means it's essentially non-deterministic. The compiler should only be loading generators / analyzers that are explicitly passed into the compiler. Hence there should be no probing here, instead every assembly name should be checked against a pre-defined list of potential assemblies. If they name isn't there then we should be falling back to the default context.

Reply:
Thanks for the feedback. I'll consider this part resolved now.